pintc/src/predicate/analyse/type_check.rs

```diff
@@ -1987,89 +2044,97 @@ impl Contract {
+                                span: span.clone(),
+                            },
+                        });
+                        Inference::Type(Type::Error(span.clone()))
+                    }
 
                     TupleAccess::Index(idx) => {
                         if let Some(field_ty) = tuple_ty.get_tuple_field_type_by_idx(*idx) {
-                            Ok(Inference::Type(field_ty.clone()))
+                            Inference::Type(field_ty.clone())
                         } else {
-                            Err(Error::Compile {
+                            handler.emit_err(Error::Compile {
                                 error: CompileError::InvalidTupleAccessor {
                                     accessor: idx.to_string(),
                                     tuple_type: self.with_ctrct(tuple_ty).to_string(),
                                     span: span.clone(),
                                 },
-                            })
+                            });
+                            Inference::Type(tuple_ty.clone())
                         }
                     }
 
                     TupleAccess::Name(name) => {
                         if let Some(field_ty) = tuple_ty.get_tuple_field_type_by_name(name) {
-                            Ok(Inference::Type(field_ty.clone()))
+                            Inference::Type(field_ty.clone())
                         } else {
-                            Err(Error::Compile {
+                            handler.emit_err(Error::Compile {
                                 error: CompileError::InvalidTupleAccessor {
                                     accessor: name.name.clone(),
                                     tuple_type: self.with_ctrct(&tuple_ty).to_string(),
                                     span: span.clone(),
                                 },
-                            })
+                            });
+                            Inference::Type(tuple_ty.clone())
                         }
                     }
                 }
             } else {
-                Err(Error::Compile {
+                handler.emit_err(Error::Compile {
                     error: CompileError::TupleAccessNonTuple {
                         non_tuple_type: self.with_ctrct(tuple_ty).to_string(),
                         span: span.clone(),
                     },
-                })
+                });
+                Inference::Type(tuple_ty.clone())
             }
         } else {
-            Ok(Inference::Dependant(tuple_expr_key))
+            Inference::Dependant(tuple_expr_key)
         }
     }
 
     fn infer_generator_expr(
         &self,
+        handler: &Handler,
         kind: &GeneratorKind,
         ranges: &[(Ident, ExprKey)],
         conditions: &[ExprKey],
         body_expr_key: ExprKey,
         span: &Span,
-    ) -> Result<Inference, Error> {
+    ) -> Inference {
```

Comment:
So I would really prefer if all these methods return `Result<Inference, ErrorEmitted>` instead of just `Inference`.

Reply:
Good catch! I was inferring the wrong type in array accesses. This also helped remove even more cascading errors in other tests.